packages/expo-linking/src/Linking.ts

```diff
@@ -0,0 +1,338 @@
+    );
+  }
+
+  let hostUri = getHostUri() || '';
+  if (usesCustomScheme() && isExpoHosted()) {
+    hostUri = '';
+  }
+
+  if (path) {
+    if (isExpoHosted() && hostUri) {
+      path = `/--/${removeLeadingSlash(path)}`;
+    }
+
+    if (!path.startsWith('/')) {
+      path = `/${path}`;
+    }
+  } else {
+    path = '';
+  }
+
+  // merge user-provided query params with any that were already in the hostUri
+  // e.g. release-channel
+  let queryString = '';
+  const queryStringMatchResult = hostUri.match(/(.*)\?(.+)/);
+  if (queryStringMatchResult) {
+    hostUri = queryStringMatchResult[1];
+    queryString = queryStringMatchResult[2];
+    let paramsFromHostUri = {};
+    try {
+      const parsedParams = qs.parse(queryString);
+      if (typeof parsedParams === 'object') {
+        paramsFromHostUri = parsedParams;
+      }
+    } catch (e) {}
+    queryParams = {
+      ...queryParams,
+      ...paramsFromHostUri,
+    };
+  }
+  queryString = qs.stringify(queryParams);
+  if (queryString) {
+    queryString = `?${queryString}`;
+  }
+
+  hostUri = removeTrailingSlash(hostUri);
+
+  return encodeURI(`${scheme}://${hostUri}${path}${queryString}`);
+}
+
+/**
+ * Returns the components and query parameters for a given URL.
+ *
+ * @param url Input URL to parse
+ */
+export function parse(url: string): ParsedURL {
+  validateURL(url);
+
+  const parsed = URL(url, /* parseQueryString */ true);
+
+  for (const param in parsed.query) {
+    parsed.query[param] = decodeURIComponent(parsed.query[param]!);
+  }
+  const queryParams = parsed.query;
+
+  const hostUri = getHostUri() || '';
+  const hostUriStripped = removePort(removeTrailingSlashAndQueryString(hostUri));
+
+  let path = parsed.pathname || null;
+  let hostname = parsed.hostname || null;
+  let scheme = parsed.protocol || null;
+
+  if (scheme) {
+    // Remove colon at end
+    scheme = scheme.substring(0, scheme.length - 1);
+  }
+
+  if (path) {
+    path = removeLeadingSlash(path);
+
+    let expoPrefix: string | null = null;
+    if (hostUriStripped) {
+      const parts = hostUriStripped.split('/');
+      expoPrefix = parts
+        .slice(1)
+        .concat(['--/'])
+        .join('/');
+    }
+
+    if (isExpoHosted() && !usesCustomScheme() && expoPrefix && path.startsWith(expoPrefix)) {
+      path = path.substring(expoPrefix.length);
+      hostname = null;
+    } else if (path.indexOf('+') > -1) {
+      path = path.substring(path.indexOf('+') + 1);
+    }
+  }
+
+  return {
+    hostname,
+    path,
+    queryParams,
+    scheme,
+  };
+}
+
+/**
+ * Add a handler to Linking changes by listening to the `url` event type
+ * and providing the handler
+ *
+ * See https://facebook.github.io/react-native/docs/linking.html#addeventlistener
```

Comment:
If we documented the events here, it'd make the docs easier to read and the API more stable. Also, instead of `addEventListener` we should expose `addURLChangeListener`(for example, that's just a suggestion) that doesn't take an event type argument. This makes it more straightforward to statically type the listener and lends itself to more static and more readable application code.